lisa/trace.py

```diff
@@ -294,7 +314,812 @@ def get_view(self, window):
+
+class _AvailableTraceEventsSet(Set):
+    """
+    Smart set that uses demand event loading on the trace to check whether an
+    event is present or not.
+    """
+    def __init__(self, trace):
+        self._trace = trace
+
+    def __contains__(self, event):
+
+        def check_event(event, raw):
+            # Try to parse the event in case it was not parsed already
+            with contextlib.suppress(MissingTraceEventError):
+                self._trace.df_events(event=event, raw=raw)
+
+            return self._trace._parsed_events[event]
+
+        try:
+            return self._trace._parsed_events[event]
+        except KeyError:
+            # Check for raw first, as it's more likely to already be parsed
+            try:
+                return check_event(event, raw=True)
+            # not there yet ? Maybe it's sometimes only available as non-raw,
+            # like cpu_frequency
+            except KeyError:
+                return check_event(event, raw=None)
+
+    @property
+    def _available_events(self):
+        return {
+            event
+            for event, available in self._trace._parsed_events.items()
+            if available
+        }
+
+    def __iter__(self):
+        return iter(self._available_events)
+
+    def __len__(self):
+        return len(self._available_events)
+
+    def __str__(self):
+        return str(self._available_events)
+
+
+class PandasDataDesc(Mapping):
+    """
+    Pandas data descriptor.
+
+    :param spec: Specification of the data as a key/value mapping.
+
+    This holds all the information needed to uniquely identify a
+    :class:`pandas.DataFrame` or :class:`pandas.Series`. It is used to manage
+    the cache and swap.
+
+    It implements the :class:`collections.abc.Mapping` interface, so
+    specification keys can be accessed directly like from a dict.
+
+    .. note:: Once introduced in a container, instances must not be modified,
+        directly or indirectly.
+
+    :ivar normal_form: Normal form of the descriptor. Equality is implemented
+        by comparing this attribute.
+    :vartype normal_form: PandasDataDescNF
+    """
+
+    def __init__(self, spec):
+        self.spec = spec
+        self.normal_form = PandasDataDescNF(self.__dict__)
+
+    def __getitem__(self, key):
+        return self.spec[key]
+
+    def __iter__(self):
+        return iter(self.spec)
+
+    def __len__(self):
+        return len(self.spec)
+
+    @classmethod
+    def from_kwargs(cls, **kwargs):
+        """
+        Build a :class:`PandasDataDesc` with the specifications as keyword
+        arguments.
+        """
+        return cls(spec=kwargs)
+
+    def __repr__(self):
+        return '{}({})'.format(
+            self.__class__.__name__,
+            ', '.join(
+                '{}={!r}'.format(key, val)
+                for key, val in self.__dict__.items()
+            )
+        )
+
+    def __eq__(self, other):
+        return self.normal_form == other.normal_form
```

Comment:
I was thinking about that but then I realized that the only place where `__eq__` is used is by containers like dict, which precisely will use `hash()` before calling `__eq__`. In that case, this would add a redundant call to `__hash__`.

Reply:
Where could I find this info? I stumbled on https://hg.python.org/cpython/file/52f68c95e025/Objects/tupleobject.c#l541 which didn't seem to use a hash

Reply:
```
>>> {[1]: 2}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```
That also hold for sets: keys must be hashable, since hashtables are used. Assuming no collision, the lookup will use the hash and then the comparison to make sure it found the right entry. Since collisions are somewhat uncommon, there is no point in running hash() again in `__eq__`, since the fact that `__eq__` is called means the hash was the same.

The link you point out is for tuples, which don't require their content to be hashable, so they don't use hash. My point is that the only container in which `PandasDataDescNF` end up are dict keys. There isn't really any other place where we compare these stuff, since their raison-d'être is to be used as swap index.

Reply:
Right, I think I went a level too far. I was focused on the internal representation (`PandasDataDescNF._nf`) which *is* a tuple, but it isn't being used here (we just compare the containers of the internal representation)